streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -1185,7 +1262,26 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
 
             // do not have serde for joined result;
             // also for key serde we do not inherit from either since we cannot tell if these two serdes are different
-            return new KStreamImpl<>(joinMergeName, joined.keySerde(), null, allSourceNodes, false, joinGraphNode, builder);
+            return new KStreamImpl<>(joinMergeName, streamJoinedInternal.keySerde(), null, allSourceNodes, false, joinGraphNode, builder);
+        }
+
+        private void assertWindowSettings(final WindowBytesStoreSupplier supplier, final JoinWindows joinWindows) {
+            final long joinGrace = joinWindows.gracePeriodMs() < 0 ? 0 : joinWindows.gracePeriodMs();
+            final boolean allMatch = supplier.retentionPeriod() == (joinWindows.size() + joinGrace) &&
+                supplier.windowSize() == joinWindows.size() && supplier.retainDuplicates();
+            if (!allMatch) {
+                throw new StreamsException(String.format("Window settings mismatch. WindowBytesStoreSupplier settings %s must match JoinWindows settings %s", supplier, joinWindows));
```

Comment:
This is fine, but it's kind of a usability pain... Do you think it would be possible at this point just to override the supplier's settings with the ones from JoinWindows and maybe log a warning instead? Or should we just not worry about it right now?

Reply:
We could be we don't know the underlying type of the `supplier` i.e  `Stores.persitentWindowStore` or `Stores.inMemoryWindowStore` or a custom supplier.  So IMHO this needs to stay as is.